jadx-gui/src/main/java/jadx/gui/device/protocol/ADB.java

```diff
@@ -73,29 +70,21 @@ private static boolean execCommandAsync(OutputStream outputStream,
 		return null;
 	}
 
-	private static byte[] readServiceProtocol(InputStream stream) {
-		byte[] bytes = null;
-		byte[] buf = new byte[4];
+	static byte[] readServiceProtocol(InputStream stream) {
 		try {
-			int len = stream.read(buf, 0, 4);
-			if (len == 4) {
-				len = unhex(buf);
-				if (len == 0) {
-					return new byte[0];
-				}
-				if (len != -1) {
-					buf = new byte[len];
-					if (stream.read(buf, 0, len) == len) {
-						bytes = buf;
-					}
-				}
+			byte[] buf = IOUtils.readNBytes(stream, 4);
+			int len = unhex(buf);
+			if (len == 0) {
+				return new byte[0];
 			}
-		} catch (IOException ignore) {
+			return IOUtils.readNBytes(stream, len);
+		} catch (IOException e) {
+			LOG.error("Failed to read readServiceProtocol: {}", e.toString());
+			return new byte[0];
```

Comment:
```suggestion
			return null;
```
This fix endless `Failed to read readServiceProtocol: java.net.SocketException: Socket closed` error